### PR TITLE
qtgui:Added Time Raster Sink Control Axis Configuration

### PR DESCRIPTION
**Pull request by ghostop14**
This update allows a time raster sink's x-axis minimum, maximum,
and label to now be specified in the block configuration options.
This allows the block to optionally be formatted to match a waterfall plot
in terms of having frequency on the lower axis and on hovers.

### CODE REVIEW

**Comment by ghostop14** [comment]
@mormj waiting for the final log, but it looked like it may be getting flagged for formatting on the pybind files?  (there was one true format in a lib file I fixed).  The pybind files looked fine to me and I tried locally to clang them to see if it output anything different and they seemed to compare the same to the original.  Just a bit new to the pybind stuff, so it may be something simple on my end.  Thx!

**Comment by mormj** [comment]
The CI clang-format seems to want a very slight variation.  Not sure why local clang-format wouldn't have done the same thing.
```
 -        .def("pyqwidget",
-             [](std::shared_ptr<time_raster_sink_b> p) {
-                 return PyLong_AsLongLong(p->pyqwidget());
-             },
-             D(time_raster_sink_b, pyqwidget))
+        .def(
+            "pyqwidget",
+            [](std::shared_ptr<time_raster_sink_b> p) {
+                return PyLong_AsLongLong(p->pyqwidget());
+            },
+            D(time_raster_sink_b, pyqwidget))
```

**Comment by ghostop14** [comment]
Looks like it wants the indent pulled back one space.  But I didn't even touch that block of code, I only added a couple new items at the end.  And if I make that change it doesn't align with the rest of the file.  Weird for sure.  

Just check my thought process here (run on Ubuntu 18.04):
clang-format time_raster_sink_b_python.cc > test.cc
diff -s time_raster_sink_b_python.cc test.cc
Files time_raster_sink_b_python.cc and test.cc are identical

I'm a little afraid to make the indent change on the one item since it won't match the rest of the file?

**Comment by mormj** [comment]
clang-format want what clang-format wants :)

```
josh@josh-dev:/share/gnuradio/grdev/src/gnuradio$ find ./ -iname *.h | xargs clang-format -i
josh@josh-dev:/share/gnuradio/grdev/src/gnuradio$ find ./ -iname *.cc | xargs clang-format -i
josh@josh-dev:/share/gnuradio/grdev/src/gnuradio$ git diff
```
If I run the above on your patch, the git diff matches what the CI produced

**Comment by ghostop14** [comment]
lol, so running that, it changed some of the others in the bindings directory that I never touched, but not the 2 time_raster_sink files.  Now it's just weird.  (Yes I sync'd with master before I made any original edits).  I'll remove the single space (copy/paste the edits it produced) and see if it complains.

	modified:   ber_sink_b_python.cc
	modified:   const_sink_c_python.cc
	modified:   edit_box_msg_python.cc
	modified:   eye_sink_c_python.cc
	modified:   eye_sink_f_python.cc
	modified:   freq_sink_c_python.cc
	modified:   freq_sink_f_python.cc
	modified:   histogram_sink_f_python.cc
	modified:   number_sink_python.cc
	modified:   sink_c_python.cc
	modified:   sink_f_python.cc
	modified:   time_sink_c_python.cc
	modified:   time_sink_f_python.cc
	modified:   vector_sink_f_python.cc
	modified:   waterfall_sink_c_python.cc
	modified:   waterfall_sink_f_python.cc

**Comment by ghostop14** [comment]
I just fresh cloned the PR branch on another system and reran clang-format.  Same result.  I copied out the log output from the test.  The formatting failure says there's 15 leading spaces on those lines, but there's only 14.  I even checked for tabs.  Nothing.

**Comment by mormj** [comment]
@ghostop14 - that is very strange.  clearly clang-format is behaving differently on your system than on the CI servers and on my machine, and this is something that we need to understand or I suspect there will be more divergence down the line.  Perhaps something is ambiguous in the .clang-format settings

I have Ubuntu 20.04 with 
```
$ clang-format --version
clang-format version 10.0.0-4ubuntu1 
```

**Comment by mormj** [comment]
@ghostop14 - updated the clang format on those two files here: https://github.com/mormj/gnuradio/tree/ghostop14-raster_fix 

Tested it out an it works well.  Couple of questions and perhaps other improvements to the raster sink:
- Any reason to not have a configurable Y axis label
- The "Line 1 label" on the config tab doesn't seem to do anything
- Might be nice to have the same configurability on the QT GUI Time Sink

**Comment by ghostop14** [comment]
@mormj, thanks!  Got the 2 files.  Yeah no reason I can't add y-axis config too.  I also noticed that the hover y-axis value runs the opposite way than the axis itself but I didn't want to disrupt the "status quo".  I was just focused on the x-axis for mirroring the waterfall capability.  I'll add in the same for the y-axis and I'll take a look at that label issue.  I can also incorporate the same capabilities into the time sink while I'm in there and just combine them in this PR.

FYI, that clang thing is weird.  When I open those 2 new files up in an editor, it clearly looks like that first one it was complaining about is unindented 1 space more than the others.  But if it passes the check, I'll go with it :)

**Comment by ghostop14** [comment]
@mormj I got the y-axis config integrated.  I still want to work on the time series one tonight, but I wanted to get some feedback on a couple items.
1. I looked at that series label not being used, and that config tab behaves just like the standard waterfall plot does.  Entering labels in each don't seem to do anything, and the color settings for an individual series seem to apply across all series (changing to white hot rather than multicolor on series 2 changes the entire color).  I'd table updating that to a more systemic fix across both waterfall and the raster after we sort out what it should even do?  I'm not sure labels, or even multiple inputs make a lot of sense on these plots, but I don't know if anyone is using that functionality for something creative.  If it's not really used, we may just want to remove that series label field, and make the color choice be a global rather than per-series setting.  
2. Take a look in TimeRasterPlot.cc on some of the %.0f, %0.1f, %0.2e print format choices around line 113.  I'm not sure I like the all-in either %.2e or %.0f so I tried to make some reasonable decisions about displayed precision but I may be over-thinking it.  Any suggestions?

Thanks!

**Comment by ghostop14** [comment]
@mormj, one more note on the clang issue (this last push was just try-and-adjust till it got accepted).

I'm on Ubuntu 18.04.5 LTS fully updated.
clang-format --version
clang-format version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)

I upgraded to 10.0.0 to match the prebuilt 18.04 version available, and still odd.  Below (top) is what I had to change time_raster_sink_b_python.cc to to get the PR check to pass, the bottom is what clang-format 10 produced (same as 6 produced):

clang-format time_raster_sink_b_python.cc > test.cc
diff time_raster_sink_b_python.cc test.cc
<         .def(
<             "pyqwidget",
<             [](std::shared_ptr<time_raster_sink_b> p) {
<                 return PyLong_AsLongLong(p->pyqwidget());
<             },
<             D(time_raster_sink_b, pyqwidget))
---
>         .def("pyqwidget",
>              [](std::shared_ptr<time_raster_sink_b> p) {
>                  return PyLong_AsLongLong(p->pyqwidget());
>              },
>              D(time_raster_sink_b, pyqwidget))

**Comment by ghostop14** [comment]
I also looked at the QT GUI Time Sink, and looks like the y axis is configurable.  Since it's a time sink, I think the time on the x-axis is appropriate, so I don't know that I would change it?

I'd stick with this PR I think as-is (pending more feedback / comments of course)

File: gr-qtgui/grc/qtgui_time_raster_x.block.yml [hunk 1]
```diff
@@ -31,6 +31,52 @@ parameters:
     label: Num. Cols
     dtype: int
     default: 256
+<<<<<<< HEAD
```
Comment (mormj) [inline gr-qtgui/grc/qtgui_time_raster_x.block.yml:34]: Looks like some merge conflicts snuck in here

Comment (ghostop14) [reply gr-qtgui/grc/qtgui_time_raster_x.block.yml]: Ugh.  Git and I weren't agreeing the other day.  I thought I fixed all of those.  Just pushed an update.

File: gr-qtgui/include/gnuradio/qtgui/TimeRasterDisplayPlot.h [hunk 2]
```diff
@@ -40,8 +40,17 @@ class TimeRasterDisplayPlot : public DisplayPlot
                    setColorMapTitleFontSize)
 
 public:
-    TimeRasterDisplayPlot(
-        int nplots, double samp_rate, double rows, double cols, QWidget*);
+    TimeRasterDisplayPlot(int nplots,
```
Comment (willcode) [inline gr-qtgui/include/gnuradio/qtgui/TimeRasterDisplayPlot.h:43]: These signatures get a little wild. Is there a way to use X & Y structs? Does that translate to Python well?

File: gr-qtgui/lib/TimeRasterDisplayPlot.cc [hunk 3]
```diff
@@ -38,15 +38,38 @@ namespace pt = boost::posix_time;
...
 public:
-    QwtXScaleDraw() : QwtScaleDraw(), TimeScaleData() {}
+    QwtXScaleDraw(double start_value = 0.0, double end_value = 0.0, int max_scale = 1024)
+        : QwtScaleDraw(),
+          TimeScaleData(),
+          d_start_value(start_value),
+          d_end_value(end_value),
+          d_max_scale(max_scale)
+    {
+        d_delta_value = d_end_value - d_start_value;
```
Comment (willcode) [inline gr-qtgui/lib/TimeRasterDisplayPlot.cc:55]: Autoscales on any start_value == end_value. Not wrong, but I can picture the user who puts 10.0 in for both and has no clue what happened.

File: gr-qtgui/lib/TimeRasterDisplayPlot.cc [hunk 4]
```diff
@@ -38,15 +38,38 @@ namespace pt = boost::posix_time;
...
-        return QwtText(QString::number(secs, 'f', 2));
+        if (d_start_value == d_end_value) {
+            // no scale was provided.  Default to seconds.
+            double secs = double(value * getSecondsPerLine());
+            return QwtText(QString::number(secs, 'f', 2));
+        } else {
+            // User-defined scale provided.
+            double x_label =
+                d_start_value + (double)value / (double)d_max_scale * d_delta_value;
+            return QwtText(QString("").sprintf("%.2e", x_label));
```
Comment (willcode) [inline gr-qtgui/lib/TimeRasterDisplayPlot.cc:71]: What happens with "%.2e" when X is large numbers, e.g., a frequency?

File: gr-qtgui/lib/TimeRasterDisplayPlot.cc [hunk 5]
```diff
@@ -63,15 +86,36 @@ class QwtXScaleDraw : public QwtScaleDraw, public TimeScaleData
...
+            // no scale was provided.  Default to row number.
+            return QwtText(QString("").sprintf("%.0f", value));
+        } else {
+            // User-defined scale provided.
+            double y_label =
+                d_start_value + (double)value / (double)d_rows * d_delta_value;
+
+            // Just add some reasonable boundary on going to e notation rather than always
+            // or not.
+            if ((d_delta_value > 999.0) or (d_delta_value <= 1.0)) {
```
Comment (willcode) [inline gr-qtgui/lib/TimeRasterDisplayPlot.cc:113]: Why different labeling for X and Y?

File: gr-qtgui/lib/TimeRasterDisplayPlot.cc [hunk 6]
```diff
@@ -148,27 +212,73 @@ class TimeRasterZoomer : public QwtPlotZoomer,
...
+        } else {
+            // Hover based on user-defined scale
+            double x = dp.x();
+            double y = dp.y();
+            if (d_y_start_value != d_y_end_value) {
+                y = d_y_start_value + y / (double)d_rows * d_y_delta_value;
+            }
+
+            double x_label = d_x_start_value + x / (double)d_cols * d_x_delta_value;
+            if ((d_y_delta_value > 999.0) or (d_y_delta_value <= 1.0)) {
```
Comment (willcode) [inline gr-qtgui/lib/TimeRasterDisplayPlot.cc:239]: Does this work for all the ranges users are likely to use?

**Review by willcode — COMMENTED** [review]
No show stoppers. Built and ran one quick flowgraph. There are a lot of parameters now! Wondering if the axis labels will work in some situations where min and max are both large numbers.

File: gr-qtgui/include/gnuradio/qtgui/TimeRasterDisplayPlot.h (hunk 2 above)
Comment (willcode) [reply gr-qtgui/include/gnuradio/qtgui/TimeRasterDisplayPlot.h:43]: What about set_x_axis(), set_x_label(), etc., similar to the way time_sink and others work?

**Comment by 777arc** [comment]
I added a note about these new params to the block doc, in anticipation of this being merged at some point, they are important params to be able to tweak, I bet I would have used this block more in the past if they were tweakable.

File: gr-qtgui/grc/qtgui_time_raster_x.block.yml [hunk 7]
```diff
@@ -31,6 +31,36 @@ parameters:
...
+    hide: part
+-   id: y_start_value
+    label: Y-Axis Start Value
+    dtype: float
+    default: 0.0
+    hide: part
+-   id: y_end_value
+    label: Y-Axis End Value
+    dtype: float
+    default: 0.1*200
```
[the remark this answers is not in the document]
Comment (ghostop14) [reply gr-qtgui/grc/qtgui_time_raster_x.block.yml:62]: Hi willcode, just getting back to this PR.  I'll push an update later tonight and work through the comments.  Thanks!

File: gr-qtgui/lib/TimeRasterDisplayPlot.cc (hunk 4 above)
Comment (ghostop14) [reply gr-qtgui/lib/TimeRasterDisplayPlot.cc:71]: I mirrored one of the other blocks to keep it similar, but I'm with you.  I'm not a fan of the fixed %.2e.  Frequencies come out as e+9 kind of numbers.  I'm going to play with some options tonight and see if I can put together a better solution.

**Comment by ghostop14** [comment]
@willcode Just pushed some updates.  Removed the 0.1*200  in the yml, and changed the scale labels to an adaptive scale based on power-of-10 like the time sink's approach. I think I like this better.

**Comment by willcode** [comment]
@ghostop14 what do you think about using set_x_axis(), set_x_label(), ..., same as time_sink, instead of adding all the new params to the constructors? That leaves the API a little cleaner.

**Comment by ghostop14** [comment]
@willcode, I know it's a lot of constructor params, but all of them seem to follow that format.  This keeps it backward-compatible.  

If GR 3.9 is heading towards release, it would be great to get this merged in.

**Comment by willcode** [comment]
@ghostop14 The graphical sinks vary quite a bit wrt to signature/setters. Personally, I like the ones that minimize params and use setters instead (e.g., Constellation Sink, Eye Sink, Frequency Sink, Time Raster Sink, Time Sink, Vector Sink, Waterfall Sink).

Any other opinions? If this tends to be used only via GRC, would it be reasonable to merge as-is and change the signature when we get time?

**Comment by willcode** [comment]
@ghostop14 I'll do the mods I'm thinking of and see if they work for you.

**Comment by ghostop14** [comment]
@willcode sounds good!

**Comment by willcode** [comment]
@ghostop14 take a look at https://github.com/willcode/gnuradio/tree/raster_fix. This is rebased on master, too. I don't think the scaling works right (in your code either?) where the unit is scaled to other than a multiple of 10**3. The lower levels of code could still use a little work, but if the user-facing signatures are good then all that can be fixed later if needed.

**Comment by mbr0wn** [comment]
@ghostop14 Can you please rebase this? It has conflicts now.

@willcode If you're happy with the changeset, can you please approve the PR? It helps us see what's looking good.

I see this PR as low-risk and would like to get it merged. I'm not sure if @marcusmueller wants to put it onto the maint-3.9 branch before we tag the release, but it should go onto master soon.

**Comment by willcode** [comment]
@mbr0wn @ghostop14 the raster_fix branch in my repo is rebased and does things more like the other sinks. There is still a bug where incorrect number scaling is done. I can fix that and submit a new join PR, or @ghostop14 can incorporate. Either way works for me.

File: gr-qtgui/lib/TimeRasterDisplayPlot.cc [hunk 8]
```diff
@@ -38,15 +39,91 @@ namespace pt = boost::posix_time;
...
+            // no scale was provided.  Default to seconds.
+            double secs = double(value * getSecondsPerLine());
+            return QwtText(QString::number(secs, 'f', 2));
+        } else {
+            // User-defined scale provided.
+            double x_label =
+                d_start_value + (double)value / (double)d_max_scale * d_delta_value;
+
+            // scale for units tag
+            x_label /= d_ten_scale;
```
Comment (willcode) [inline gr-qtgui/lib/TimeRasterDisplayPlot.cc:124]: Account for selected `units3`.

File: gr-qtgui/lib/TimeRasterDisplayPlot.cc [hunk 9]
```diff
@@ -63,15 +140,80 @@ class QwtXScaleDraw : public QwtScaleDraw, public TimeScaleData
...
+        if (d_start_value == d_end_value) {
+            // no scale was provided.  Default to row number.
+            return QwtText(QString("").sprintf("%.0f", value));
+        } else {
+            // User-defined scale provided.
+            double y_label =
+                d_start_value + (double)value / (double)d_rows * d_delta_value;
+
+            // scale for units tag
+            y_label /= d_ten_scale;
```
Comment (willcode) [inline gr-qtgui/lib/TimeRasterDisplayPlot.cc:214]: Account for selected `units3`.

**Comment by willcode** [comment]
@mbr0wn @ghostop14 I put in https://github.com/gnuradio/gnuradio/pull/4092 with the scale fix, rebase.

**Comment by mbr0wn** [comment]
Closing in favour of #4092. @ghostop14 I hope you're OK with that process! Please let us know if not. We can always reopen this.

**Comment by ghostop14** [comment]
Hey @mbr0wn !  Sure thing, I'm good with it.  Sorry, been a busy week.  Thanks @willcode!

**Pull request closed without merging**